From 7826585128c0e22afde1c392027efaa703f75fbb Mon Sep 17 00:00:00 2001
From: Dmitry Karikh <the.dr.hax@gmail.com>
Date: Sun, 20 Aug 2017 06:58:34 +0300
Subject: [PATCH 039/100] HACK: drm: i915: Turn off display completely when
 brightness is set to 0

This hack is needed by Android to turn display off when device is locked.
Without it screen just switches to 0 brightness and still glowing in dark.

Tested on Asus Transformer Book T100TAM, Android 7.1.2

UPDATE 2019-09-01: Compatibility with Linux v5.2+.
---
 drivers/gpu/drm/i915/display/intel_panel.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/gpu/drm/i915/display/intel_panel.c b/drivers/gpu/drm/i915/display/intel_panel.c
index 5fdf52643150..67698125d7c8 100644
--- a/drivers/gpu/drm/i915/display/intel_panel.c
+++ b/drivers/gpu/drm/i915/display/intel_panel.c
@@ -1287,6 +1287,8 @@ static void intel_panel_set_backlight(const struct drm_connector_state *conn_sta
 				      u32 user_level, u32 user_max)
 {
 	struct intel_connector *connector = to_intel_connector(conn_state->connector);
+	struct intel_crtc *crtc = to_intel_crtc(conn_state->crtc);
+	struct intel_crtc_state *crtc_state = to_intel_crtc_state(crtc->base.state);
 	struct drm_i915_private *dev_priv = to_i915(connector->base.dev);
 	struct intel_panel *panel = &connector->panel;
 	u32 hw_level;
@@ -1294,6 +1296,13 @@ static void intel_panel_set_backlight(const struct drm_connector_state *conn_sta
 	if (!panel->backlight.present)
 		return;
 
+	// HACK: Turn off display completely when brightness is set to 0
+	if (user_level == 0 && panel->backlight.enabled) {
+		intel_panel_disable_backlight(conn_state);
+	} else if (!panel->backlight.enabled) {
+		intel_panel_enable_backlight(crtc_state, conn_state);
+	}
+
 	mutex_lock(&dev_priv->backlight_lock);
 
 	drm_WARN_ON(&dev_priv->drm, panel->backlight.max == 0);
-- 
2.17.1

